lib/dao.js

```diff
@@ -424,13 +435,26 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
 
         var connector = self.getConnector();
 
+        context = {
+          Model: Model,
+          where: ctx.where,
+          data: removeUndefined(data),
+          currentInstance: inst,
+          hookState: ctx.hookState,
+          options: options
+        };
+
         if (Model.settings.validateUpsert === false) {
-          update = removeUndefined(update);
-          if (connector.updateOrCreate.length === 4) {
-            connector.updateOrCreate(Model.modelName, update, options, done);
-          } else {
-            connector.updateOrCreate(Model.modelName, update, done);
-          }
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return cb(err);
+
+            update = removeUndefined(update);
+            if (connector.updateOrCreate.length === 4) {
+              connector.updateOrCreate(Model.modelName, update, options, done);
+            } else {
+              connector.updateOrCreate(Model.modelName, update, done);
+            }
+          });
```

Comment:
The original code contained some duplication. With your changes in place, I feel the amount of duplicated code became unmanageable. Could you please refactor (extract) a new method with the shared code and use it in both places?

Example:

```js
function callConnector() {
  update = removeUndefined(update);
  context = {
    Model: Model,
    where: ctx.where,
    data: update,
    currentInstance: inst,
    hookState: ctx.hookState,
    options: options
  };
  Model.notifyObserversOf('persist', context, function(err) {
    if (err) return done(err);
    if (connector.updateOrCreate.length === 4) {
      connector.updateOrCreate(Model.modelName, update, options, done);
    } else {
      connector.updateOrCreate(Model.modelName, update, done);
    }
  });
}